### DISCUSSION TITLE
Losing cell format when writing from reader (Ods)

### DISCUSSION THREAD

**Discussion by ThomasBoom89** [PHPOffice/PhpSpreadsheet#2938] [Q&A] [answered]
Hello,

would like to know if the behaviour of following example is desired or if it is a bug.

I have an ods file with a date formatted cell. I load this file with the ods reader and then add a value to another cell (doesn't matter). Now i construct a writer and would like to save the data. If i open the saved file with an office suite, the date formatted cell is now a floating point number in a float typed field (also looked directly into the xml). I did a deep dive into the reader and writer. The writer has limitation about what type can be processed to the file. The date type can not be processed, same as with time etc.. The reader converts the date type to a float and sets the type to number type ('n'). With this in mind, this seems not to be a bug but also, for me, not the expected behaviour. Can you confirm my thoughts about it? And if yes, is this just lack of implementation or has this a technical reason?

Example:
```php
$reader      = new OdsFileReader();
$spreadsheet = $reader->load($filePath);
// do some changes to worksheet
$writer      = new OdsFileWriter($spreadsheet);
$writer->save('path');
```

Thanks

**Comment by MarkBaker** [comment, marked as answer]
Dates/Times are always a MS Serialized  Timestamp value internally; and the Ods reader reads and interprets this correctly, converting it to the correct numeric representation for that date. What isn't fully implemented is reading the format mask for the value (the code just uses a generic date/time format mask when reading), or writing with the correct format mask

**Reply by ThomasBoom89** [reply]
Thank you for your quick response.

**Comment by abivia** [comment]
Any idea what the scope of this work is? (And if it's large is there any chance this is on a roadmap?) ODS is becoming more popular and this limitation means either exporting cells as formatted strings or confusing consumers of the generated files, both are sub-optimal solutions.

**Comment by oleibman** [comment]
There were a number of PRs merged in February and March of this year to improve style support for Ods.

**Comment by abivia** [comment]
I'm running v5.9.0 with code resembling
```
// Some other style settings here
$style['numberFormat'] = ['formatCode' => 'yyyy-mm-dd'];
$sheet->getStyle([$col, $row])->applyFromArray($style);
```
I've tired this, various other formats, forced uppercase... and the ODS always has the date/time values as a floating point. Meanwhile XLS/HTML output works with every format I've tried. Hopefully I'm just doing it wrong...

If I'm doing it right, can you point me at the PRs? I'll try to find time to dig in with some test cases and see if I can figure out what the problem is.

**Comment by oleibman** [comment]
I would need to see more of your code to see why it is not working. I have adapted your code, and the result seems okay to me.
```php
require __DIR__ . '/PhpSpreadsheet' . '/vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Writer\Ods as OdsWriter;

class issue2938 {
    public static function testDate(): void
    {
        $spreadsheet = new Spreadsheet();
        $sheet = $spreadsheet->getActiveSheet();
        $sheet->getCell('A1')->setValue('=TODAY()');
        $col = 1;
        $row = 1;
        $style = [];
        $style['numberFormat'] = ['formatCode' => 'yyyy-mm-dd'];
        $sheet->getStyle([$col, $row])->applyFromArray($style);
        $outfile = 'issue.2938.ods';
        $writer = new OdsWriter($spreadsheet);
        $writer->save($outfile);
        echo "saved $outfile\n";
    }
}

issue2938::testDate();
```
Result:
<img width="316" height="283" alt="image" src="https://github.com/user-attachments/assets/b6b44956-582c-4312-a62b-836afdac36b4" />

**Comment by oleibman** [comment]
And, just to demonstrate the reader:
```php
    public static function readDate(): void
    {
        $infile = 'issue.2938.ods';
        $reader = new OdsReader();
        $spreadsheet = $reader->load($infile);
        $sheet = $spreadsheet->getActiveSheet();
        var_dump($sheet->getCell('A1')->getFormattedValue());
    }
```
Result:
```
C:\git\issue.2938.php:32:
string(10) "2026-08-10"
```

**Comment by abivia** [comment]
Well you can see all of the code at https://github.com/instancezero/NovelWriterExtract/blob/main/src/Lib/NovelWriter/ExtractGrid.php but it's grown from a test into an application and is in serious need of refactoring. There's a lot to wade through. ExtractGrid::formatCell() is where the format gets set. Either way it's pretty clearly my issue to fix...

**Comment by abivia** [comment]
Very strange. Seems to work for ISO date format as in your test but not other formats.

```
require __DIR__ . '/../vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Writer\Ods as OdsWriter;

class issue2938 {
    public static function testDate(): void
    {
        $spreadsheet = new Spreadsheet();
        $sheet = $spreadsheet->getActiveSheet();
        $sheet->getCell('A1')->setValue(46159.961516204);
        $col = 1;
        $row = 1;
        $style = [];
        $style['numberFormat'] = ['formatCode' => 'mmm d, yyyy'];
        $sheet->getStyle([$col, $row])->applyFromArray($style);
        echo $sheet->getCell([$col, $row])->getFormattedValue() . "\n";
        $outfile = __DIR__ . '/results/issue.2938.ods';
        $writer = new OdsWriter($spreadsheet);
        $writer->save($outfile);
        echo "saved $outfile\n";
    }
}

issue2938::testDate();
```
outputs
```
May 17, 2026
saved {path}
```
But the format doesn't take in the sheet:
<img width="157" height="78" alt="image" src="https://github.com/user-attachments/assets/e0a7100c-ec27-426a-b16e-7a6c98e411c4" />

Still, progress from having the column read 46159.961516204.

**Comment by oleibman** [comment]
We are not in a position to support the infinite variety of possible number formats - it is very difficult to map back and forth between how Excel and LibreOffice define their styles. This was noted in PR #4806. Supported formats are in Writer/Ods/Cell constant NUMBER_FORMAT_METHODS. However, the notes for the PR indicate that you can customize your own if they aren't taken care of. and techniques for doing that are demonstrated in Sample55_DefinedStyles (see `function WriteAdditional`). The notes also indicate that we may be amenable to adding, on request, some unsupported styles to the builtin list.

At any rate, we don't yet support any `mmm` formats. That might be fairly easy for English, but then the question of support for other languages arises. But it could be fairly easy for the user to handle this with a custom format using the desired language. And, of course, we do accept PR's for any unsupported formats that you just have to have now.

**Comment by abivia** [comment]
That makes perfect sense. Better if it's left to the user and LibreOffice Calc to manage the complexities. I'm just happy that it looks like a date now.

**Comment by oleibman** [comment]
I misspoke about `mmm` formats. We do support some, in particular `d-mmm-yyyy` (`NumberStyle::FORMAT_DATE_XLSX15_YYYY`). You can easily use `Writer\Ods\Cell\Style::formatDateXlsx15Yyyy()` as a model to implement a similar custom format such as the one you're using.

**Comment by oleibman** [comment]
PR #4960, just submitted, demonstrates how you might deal with `mmm d, yyyy`.
